src/Bicep.Core/Emit/ResourceDependencyVisitor.cs

```diff
@@ -9,37 +9,64 @@
+    /// Modules and Resources only give immediate dependencies (NOT NESTED). Given three resources/modules a,b,c, 
+    /// i.e. if a depends on b, and b depends on c,
+    /// the graph would be {a: {b}, b: {c}}
+    /// However, variables need to be evaluated for NESTED dependencies:
+    /// i.e. if a (module/resource) depends on b (variable), and b depends c(module/resource)
+    /// the graph would be {a: {c}} 
+    /// </summary>
     public class ResourceDependencyVisitor : SyntaxVisitor
     {
+        // shared data structure across all ResourceDependencyVisitors; our only "state" during traversals
+        // Instead of traversing through variables repeatedly, we pre-calculate variable-resource dependencies before we build the graph
+        private static Dictionary<VariableSymbol, ImmutableHashSet<DeclaredSymbol>> variableDependencies = new Dictionary<VariableSymbol, ImmutableHashSet<DeclaredSymbol>>();
         private readonly SemanticModel.SemanticModel model;
-        private IDictionary<DeclaredSymbol, HashSet<DeclaredSymbol>> resourceDependencies;
-        private DeclaredSymbol? currentDeclaration;
+        // resource dependencies specific to each ResourceDependencyVisitor
+        private HashSet<DeclaredSymbol> resourceDependencies;
 
-        public static ImmutableDictionary<DeclaredSymbol, ImmutableHashSet<DeclaredSymbol>> GetResourceDependencies(SemanticModel.SemanticModel model)
+        public static ImmutableDictionary<DeclaredSymbol, ImmutableHashSet<DeclaredSymbol>> GetAllResourceDependencies(SemanticModel.SemanticModel model) 
         {
-            var visitor = new ResourceDependencyVisitor(model);
-            visitor.Visit(model.Root.Syntax);
+            var allResourceDependencies = new Dictionary<DeclaredSymbol, ImmutableHashSet<DeclaredSymbol>>();
+            var declaredSymbols = model.Root.AllDeclarations.ToList(); // gets all declaredSymbols in the document (entry point)
 
-            var output = new Dictionary<DeclaredSymbol, ImmutableHashSet<DeclaredSymbol>>();
-            foreach (var kvp in visitor.resourceDependencies)
+            // we need to precompute variable dependencies to reuse them when traversing resources or modules
+            for(int i = 0; i < declaredSymbols.Count; i++) 
             {
-                if (kvp.Key is ResourceSymbol resourceSymbol)
+                if (declaredSymbols[i] is VariableSymbol variableSymbol)
                 {
-                    output[resourceSymbol] = kvp.Value.ToImmutableHashSet();
+                    variableDependencies[variableSymbol] = GetResourceDependencies(model, variableSymbol.DeclaringSyntax);
                 }
-                if (kvp.Key is ModuleSymbol moduleSymbol)
+            }
+
+            for (int i = 0; i < declaredSymbols.Count; i++)
+            {
+                switch(declaredSymbols[i])
```

Comment:
>switch [](start = 16, length = 6)

Could reduce the amount of code by doing a foreach on .Where(symbol => symbol is ModuleSymbol || symbol is ResourceSymbol) since the properties you're accessing are on a base type (DeclaredSymbol)